### PR TITLE
Add support for Morlet wavelets that are longer than the signal

### PR DESCRIPTION
**Pull request by wmvanvliet**
Found a fixme in the code, so fixed it.

I verified that:

    np.allclose(cwt(X, Ws, use_fft=True, mode='full'), cwt(X, Ws, use_fft=False, mode='full'))

@kingjr thoughts?

### CODE REVIEW

File: mne/time_frequency/tfr.py [hunk 1]
```diff
@@ -209,12 +206,12 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
...
     for i, W in enumerate(Ws):
-        if len(W) > n_times:
-            raise ValueError('At least one of the wavelets is longer than the '
-                             'signal. Use a longer signal or shorter '
-                             'wavelets.')
         if use_fft:
             fft_Ws[i] = fft(W, fsize)
+        elif mode == 'same' and len(W) > n_times:
+            warn("At least one of the wavelets is longer than the "
+                 "signal. Switching to 'full' convolution mode.")
```
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: I'd rather make this an error. If people hit use cases maybe we can add an explicit argument to pad the signal or something.

File: mne/time_frequency/tfr.py [hunk 2]
```diff
@@ -230,14 +227,17 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
...
                 sz = int(abs(W.size - n_times)) + 1
                 offset = (n_times - sz) // 2
                 this_slice = slice(offset // decim.step,
                                    (offset + sz) // decim.step)
                 if use_fft:
                     ret = _centered(ret, sz)
                 tfr[ii, this_slice] = ret[decim]
+            elif mode == 'full' and not use_fft:
+                ret = ret[(W.size // 2):len(ret) - (W.size // 2)]
+                tfr[ii, :] = ret[decim]
```
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: You are missing a `-1` somewhere, probably need a `(W.size - 1) // 2`, as this only gives the original length back for odd-length signals:
```
>>> import numpy as np
>>> x = np.zeros(10)
>>> w = np.zeros(3)
>>> y = np.convolve(x, w)
>>> y[(w.size // 2):len(y) - (w.size // 2)].size
10
>>> w = np.zeros(4)
>>> y = np.convolve(x, w)
>>> y[(w.size // 2):len(y) - (w.size // 2)].size
9
```
If true, please add a test if possible.

File: mne/time_frequency/tfr.py (hunk 1 above)
Comment (wmvanvliet) [inline mne/time_frequency/tfr.py]: Agreed. But the error should not be thrown when using FFT, as is currently being done. As far as I can tell, the current FFT implementation already pads the signal properly.

File: mne/time_frequency/tfr.py (hunk 2 above)
Comment (wmvanvliet) [inline mne/time_frequency/tfr.py]: You're right! Fixed and test added.

File: mne/time_frequency/tfr.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: So for now, `warn` in cases where `len(W) > n_times` when `use_fft=True`, and throw an error (or fix the implementation) when `use_fft=False`?

File: mne/time_frequency/tests/test_tfr.py [hunk 3]
```diff
@@ -204,18 +205,22 @@ def test_time_frequency():
...
-                continue
-            cwt(data[0, :, :], Ws, use_fft=use_fft, mode=mode)
+            cwt(data[0], Ws, use_fft=use_fft, mode=mode)
 
     # Test decim parameter checks
     assert_raises(TypeError, tfr_morlet, epochs, freqs=freqs,
                   n_cycles=n_cycles, use_fft=True, return_itc=True,
                   decim='decim')
 
+    # When convolving in time, wavelets must not be shorter than the data
```
Comment (mmagnuski) [inline mne/time_frequency/tests/test_tfr.py]: You probably mean "longer than the data"? (for example you take data that is shorter than the wavelet here)

Comment (wmvanvliet) [inline mne/time_frequency/tests/test_tfr.py]: oops... fixed :)

**Comment by wmvanvliet** [comment]
How about this?

File: mne/time_frequency/tfr.py [hunk 4]
```diff
@@ -230,14 +233,19 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
...
+            if mode == 'valid':
                 sz = int(abs(W.size - n_times)) + 1
                 offset = (n_times - sz) // 2
                 this_slice = slice(offset // decim.step,
                                    (offset + sz) // decim.step)
                 if use_fft:
                     ret = _centered(ret, sz)
                 tfr[ii, this_slice] = ret[decim]
+            elif mode == 'full' and not use_fft:
+                start = int(np.ceil(W.size / 2.))
```
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: `(W.size + 1) // 2` is cleaner

**Comment by kingjr** [comment]
LGTM, ok for merge if travis is happy

File: mne/time_frequency/tfr.py (hunk 4 above)
Comment (wmvanvliet) [inline mne/time_frequency/tfr.py]: you're right

**Comment by wmvanvliet** [comment]
Double checked this against Jan Kujala's code. Off-by-one errors are hard.

File: mne/time_frequency/tfr.py [hunk 5]
```diff
@@ -230,14 +233,19 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
...
+            if mode == 'valid':
                 sz = int(abs(W.size - n_times)) + 1
                 offset = (n_times - sz) // 2
                 this_slice = slice(offset // decim.step,
                                    (offset + sz) // decim.step)
                 if use_fft:
                     ret = _centered(ret, sz)
                 tfr[ii, this_slice] = ret[decim]
+            elif mode == 'full' and not use_fft:
+                start = (W.size + 1) // 2 - 1
```
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: This is equivalent to `(W.size - 1) // 2` :)

File: mne/time_frequency/tfr.py [hunk 6]
```diff
@@ -230,14 +233,19 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
...
+            if mode == 'valid':
                 sz = int(abs(W.size - n_times)) + 1
                 offset = (n_times - sz) // 2
                 this_slice = slice(offset // decim.step,
                                    (offset + sz) // decim.step)
                 if use_fft:
                     ret = _centered(ret, sz)
                 tfr[ii, this_slice] = ret[decim]
+            elif mode == 'full' and not use_fft:
+                start = (W.size - 1) // 2
```
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: You called this a cosmetic change?!? Think of the effeciency gains!

**Comment by agramfort** [comment]
LGTM

**Comment by Eric89GXL** [comment]
Thanks @wmvanvliet

**Pull request merged**